File: 2 kyu/Decode the Morse code, for real/solution.java
import java.util.*;

public class MorseCodeDecoder {
    public static String decodeMorse(String morseCode) {
        StringBuilder sb = new StringBuilder();
        for (String code: morseCode.split(" {1,2}"))
            sb.append("".equals(code) ? " " : MorseCode.get(code));
        return sb.toString().trim();
    }
    
    public static String decodeBitsAdvanced(String bits) {
        bits = bits.replaceAll("^0*(.*?)0*$","$1");
        int maxGroupOf1Length = Arrays.stream(bits.split("0+"))
                                      .mapToInt(String::length)
                                      .max().orElse(1);
        var stats = Arrays.stream(bits.split("(?=1)(?<=0)|(?=0)(?<=1)"))
                          .mapToInt(String::length)
                          .summaryStatistics();
                                              
        int bound = (int)Math.ceil((stats.getMin()+maxGroupOf1Length) / 2.0);
        if (maxGroupOf1Length == stats.getMin()) bound++;
        int wordbound = (int)Math.ceil((bound + stats.getMax()) / 2.0);
        wordbound = Math.max(wordbound, 5 * stats.getMin());

        return bits.replaceAll(String.format("1{%d,}", bound), "-")
                    .replaceAll(String.format("1{1,%d}", bound), ".")
                    .replaceAll(String.format("0{%d,}", wordbound), "   ")
                    .replaceAll(String.format("0{%d,}",bound), " ")
                    .replaceAll("0","");
    }
}
_______________________________________________
public class MorseCodeDecoder {
    
    public static String decodeBitsAdvanced(String bits) {

        StringBuilder sBits = new StringBuilder(bits.replaceAll("^0+|0+$", "")).append(" ");

        int oneMin = Integer.MAX_VALUE;
        int oneMax = 0;
        int zeroMin = Integer.MAX_VALUE;
        int zeroMax = 0;

        char lastChar = '~';
        int count = 0;
        for (int i = 0; i < sBits.length(); i++) {

            if (sBits.charAt(i) != lastChar) {

                if (lastChar == '0') {
                    if (count < zeroMin) zeroMin = count;
                    if (count > zeroMax) zeroMax = count;
                }

                if (lastChar == '1') {
                    if (count < oneMin) oneMin = count;
                    if (count > oneMax) oneMax = count;
                }

                lastChar = sBits.charAt(i);
                count = 1;

            } else {
                count++;
            }

        }

        StringBuilder morse = new StringBuilder();

        lastChar = '~';
        count = 0;
        for (int i = 0; i < sBits.length(); i++) {
            if (sBits.charAt(i) != lastChar) {

                if (lastChar == '0') {

                    if (zeroMin == zeroMax) {
                        if (zeroMin >= (oneMin * 6)) morse.append("  ");
                        else if (zeroMin > oneMin) morse.append(" ");
                    }
                    else if (!minCloser(count, zeroMin, oneMax)) {
                        int between = 1 + (int) Math.ceil(( zeroMin + zeroMax) / 2.0 );
                        if ( between < count && (minCloser(count, zeroMax, oneMax) || oneMax < count)) morse.append("  ");
                        else morse.append(" ");
                    }

                }

                if (lastChar == '1') {
                    if (oneMin == oneMax) {
                        if (oneMin == zeroMin || zeroMin == Integer.MAX_VALUE) morse.append(".");
                        else if (oneMin < zeroMin) morse.append(".");
                        else morse.append("-");
                    }
                    else if (minCloser(count, oneMin, oneMax)) morse.append(".");
                    else morse.append("-");
                }

                lastChar = sBits.charAt(i);
                count = 1;


            } else {
                count++;
            }
        }

        return morse.append(" ").toString();
    }


    static boolean minCloser(int value, int min, int max){

        int rangeMin = Math.abs(min - value);
        int rangeMax = Math.abs(max - value);

        return  (rangeMin < rangeMax);
    }


    public static String decodeMorse(String morseCode) {

        String[] letters = morseCode.split(" ");
        StringBuilder answer = new StringBuilder();

        for (int i = 0; i < letters.length; i++) {
            if (letters[i].equals("")) answer.append(" ");
            else answer.append(MorseCode.get(letters[i]));
        }

        return answer.toString().trim();
    }
    
}
_______________________________________________
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Scanner;

public class MorseCodeDecoder {
    private static int tu;
    private static float thresh13;
    private static float thresh37;
    
     /**
     * Given a string of ones and its following strings of zeros,
     * returns the Morse symbol (e.g. dot, dash, new-letter, new-word)
     * which these ones and zeros signify).
     * This method works even when there is some variation in the length
     * of the time unit used.
     * 
     * @param one
     * @param zero
     * @return 
     */
    private static String nextTeleFuzzy(String one, String zero) {
        String tele = nextTeleFuzzy(one);
        if ((zero.length() >= thresh13) && (zero.length() < thresh37)) tele += " ";
        else if (zero.length() >= thresh37) tele += "   ";
        return tele;
    }
    
    /**
     * Given a string of ones, returns the Morse symbol
     * (i.e. dot or dash) which these ones signify.
     * This method works even when there is some variation in the length
     * of the time unit used.
     * 
     * @param one
     * @return 
     */
    private static String nextTeleFuzzy(String one) {
        String tele = "";
        if (one.length() <= thresh13) tele += ".";
        else tele += "-";
        return tele;
    }
    
    /**
     * Given a string of bits, which may or may not begin or end with '0's,
     * and which may have some variation in the length of the time unit used,
     * returns the Morse Code translation of this message.
     * @param bits
     * @return 
     */
    public static String decodeBitsAdvanced(String bits) {
        String morse = "";
        bits = bits.replaceAll("^[0]+", "");
        bits = bits.replaceAll("[0]+$", "");
        KMeans km = new KMeans(bits, 3);
        km.converge();
        thresh13 = (km.getTimeUnit(0) + km.getTimeUnit(1)) / 2;
        thresh37 = (km.getTimeUnit(1) + km.getTimeUnit(2)) / 2;
        if (bits.length() > 5) {
            thresh13 *= 1.2;
            thresh37 *= 1.1;
        }
        String[] ones = bits.split("0+");
        String[] zeros = bits.split("1+");
        for (int i = 0; i < zeros.length - 1; i++) {
            morse += nextTeleFuzzy(ones[i], zeros[i + 1]);
        }
        if (ones[0].length() > 0) morse += nextTeleFuzzy(ones[ones.length - 1]);
        return morse;
    }
    
    /**
     * Given a string in Morse Code, returns the English translation.
     * 
     * @param morseCode
     * @return 
     */
    public static String decodeMorse(String morseCode) {
        String results = "";
        morseCode = morseCode.trim().replaceAll(" {3}", " SPACE ");
        // Here, we'd like to trim leading and trailing whitespace.
        // We also know that three spaces are used to separate words.
        // Hence, we leave a symbol that can be tokenized for our while
        // loop to recognize as a [space] character.

        Scanner sc = new Scanner(morseCode);
        while (sc.hasNext()) { 
          String nxt = sc.next();
          if (nxt.equals("SPACE")) results += " ";
          else results += MorseCode.get(nxt);
        }
        return results;
    }
    
        /**
     * The Cluster class provides data structures and methods for clusters
     * in the KMeans algorithm.
     */
    private static class Cluster implements Comparable<Cluster> {
        private float location;
        private float centroid;
        private ArrayList<Integer> currentPoints = new ArrayList<>();
        private ArrayList<Integer> previousPoints = new ArrayList<>();
        
        /**
         * Constructors
         */
        private Cluster(float loc) {
            location = loc;
        }
        
        private Cluster() {
            location = -1;
        }
        
        /**
         * Methods for claiming currentPoints and calculating centroid.
         */
        private void addPoint(int i) {
            currentPoints.add(i);
        }
        
        private boolean didChange() {
            if (previousPoints.size() != currentPoints.size()) return true;
            else return !currentPoints.equals(previousPoints);
        }
        
        private void clearPoints() {
            previousPoints = (ArrayList<Integer>) currentPoints.clone();
            currentPoints.clear();
        }
        
        /**
         * After new points have been assigned to this cluster, this method
         * calculates the new centroid of the cluster and moves the cluster
         * to that location.
         */
        private void update() {
            float sum = 0;
            for (Integer p: currentPoints) {
                sum += p;
            }
            centroid = sum / currentPoints.size();
            setLocation(centroid);
        }
                
        /**
         * Getters and Setters.
         */
        private float getLocation() { return location; }
        private float getDistance(int point) {
            return Math.abs(location - point);
        }
        private void setLocation(float loc) { location = loc; }
        
        @Override
        public int compareTo(Cluster t) {
            if (this.getLocation() > t.getLocation()) return 1;
            else if (this.getLocation() < t.getLocation()) return -1;
            else return 0;
        }
    }
    
    private static class KMeans {    
        /**
         * KMeans attributes.
         */
        private boolean converged = false;
        private final Cluster[] clusters;
        private final String[] bitCollection; // for generating frequency dist.
        private float[] timeUnits = {0, 0, 0};
        private final HashMap<Integer, Integer> dist = new HashMap<>();
        List<Integer> keys;

        public KMeans(String stream, int numClusters) {
            this.clusters = new Cluster[numClusters];
            stream = stream.replaceAll("^[0]+", ""); // remove leading 0s
            stream = stream.replaceAll("[0]+$", ""); // remove trailing 0s

            /**
             * The following if/else block populates this.bitCollection.
             */
            if (stream.length() == 0) {
                bitCollection = new String[1];
                bitCollection[0] = "";
            }
            else {
                String[] ones = stream.split("0+");
                String[] zeros = stream.split("1+");

                if (zeros.length == 0) {
                    bitCollection = new String[1];
                    bitCollection[0] = ones[0];
                }
                else {
                    bitCollection = new String[ones.length + zeros.length - 1];
                    for (int i = 0; i < ones.length - 1; i++) {
                        bitCollection[2*i] = ones[i];
                        bitCollection[2*i+1] = zeros[i+1];
                    }
                    bitCollection[bitCollection.length - 1] = ones[ones.length - 1];
                }
            }

            /**
             * The following for loop populates the this.dist HashMap.
             */
            for (int i = 0; i < bitCollection.length; i++) {
                int l = bitCollection[i].length();
                if (!dist.containsKey(l)) {
                    dist.put(l, 1);
                }
                else dist.put(l, dist.get(l) + 1);
            }
            this.keys = new ArrayList<>(dist.keySet());
                        
            if (keys.size() == 1 || keys.size() == 2) {
                timeUnits[0] = keys.get(0);
                timeUnits[1] = keys.get(0) * 3;
                timeUnits[2] = keys.get(0) * 7;
                converged = true;
            }
            else {
                Collections.sort(keys);
                initializeClusters();
            }
        }

        /**
         * Populates this.clusters with this.numClusters Cluster objects,
         * whose initial locations are from this.keys (the minimum, the
         * maximum, and the middle between the two).
         */
        private void initializeClusters() {
            clusters[0] = new Cluster(keys.get(0));
            clusters[2] = new Cluster(keys.get(keys.size() - 1));
            clusters[1] = new Cluster(
                (keys.get(keys.size() - 1) + keys.get(0)) / 2 + 1);
        }

        /**
         * Assigns cluster-labels to each length-point from the fuzzy input,
         * which is subsequently used by the clusters to re-calculate their
         * centroids and move accordingly.
         */
        public void assignToClosestCluster() {
            clear();
            for (Integer i: keys) {
                Cluster bestCluster = new Cluster();
                float closest = Float.MAX_VALUE;
                for (Cluster c: clusters) {
                    float d = c.getDistance(i);
                    if (d < closest) {
                        closest = d;
                        bestCluster = c;
                    }
                }
                for(int j = 0; j < dist.get(i); j++) {
                    bestCluster.addPoint(i);
                }
            }
        }

        /**
         * Populates this.timeUnits[] with the first, second, and third cluster
         * means, representing the average length of 1 time unit,
         * 3 time units, and 7 time units respectively.
         */
        public void calculateTimeUnits() {
            Cluster[] sortedClusters = clusters.clone();
            Arrays.sort(sortedClusters);
            timeUnits[0] = sortedClusters[0].getLocation();
            timeUnits[1] = sortedClusters[1].getLocation();
            timeUnits[2] = sortedClusters[2].getLocation();
        }

        public void clear() {
            for (Cluster c: clusters) c.clearPoints();
        }

        /**
         * Assigns the closest Cluster to each point, calculates the centroid
         * for those Clusters based off of those points, moves the Clusters
         * to their respective centroids, and repeats until assignment on the next
         * iteration is the same.
         */
        public void converge() {
            if (!converged) {
                assignToClosestCluster();
                while (!converged) {
                    update();
                    assignToClosestCluster();
                    if (!didChange()) converged = true;
                }
                calculateTimeUnits();
            }
        }

        public boolean didChange() {
            for (Cluster c: clusters) if (c.didChange()) return true;
            return false;
        }

        public void update() {
            for (Cluster c: clusters) c.update();
        }

        /**
         * Getters and Setters.
         *
         */
        public float getTimeUnit(int index) { return this.timeUnits[index]; }
    }
}
_______________________________________________
import java.util.*;
import java.util.regex.*;
import java.util.stream.*;

public class MorseCodeDecoder {
    
    private static Pattern BITS_PATTERN  = Pattern.compile("0+|1+");
    private static Map<List<Boolean>,String> CONVERT_BITS = new HashMap<List<Boolean>,String>();
    static {CONVERT_BITS.put(Arrays.asList(true, true, true),    ".");
            CONVERT_BITS.put(Arrays.asList(true, false, true),   "-");
            CONVERT_BITS.put(Arrays.asList(true, false, false),  "-");
            CONVERT_BITS.put(Arrays.asList(false, true, true),   "");
            CONVERT_BITS.put(Arrays.asList(false, false, true),  " ");
            CONVERT_BITS.put(Arrays.asList(false, false, false), "   ");
    }
    
    public static String decodeBitsAdvanced(String bits) {
        String stripBits = bits.replaceAll("^0+|0+$", "");
        if (stripBits.length() == 0) return "";
        
        Map<Integer,Integer> ones = new HashMap<Integer,Integer>(),
                             full = new HashMap<Integer,Integer>();
        Matcher m = BITS_PATTERN.matcher(stripBits);
        while (m.find()) {
            String seq = m.group();
            full.put(seq.length(), full.getOrDefault(seq.length(), 0) + 1);
            if (seq.charAt(0) == '1')
                ones.put(seq.length(), ones.getOrDefault(seq.length(), 0) + 1);
        }
        
        int mi1 = Collections.min(ones.keySet()),
            ma1 = Collections.max(ones.keySet()),
            miF = Collections.min(full.keySet()),
            maF = Collections.max(full.keySet()),
            miG = -1, maG = -1;
        int[] limits = new int[2];
        
        Set<Integer> gaps1    = IntStream.range(mi1+1, ma1).filter(i -> !ones.containsKey(i)).boxed().collect(Collectors.toSet()),
                     gapsFull = IntStream.range(miF+1, maF).filter(i -> !full.containsKey(i)).boxed().collect(Collectors.toSet()),
                     gapsSep  = new HashSet<Integer>();
        if (!gapsFull.isEmpty()) {
            miG = Collections.min(gapsFull);
            maG = Collections.max(gapsFull);
            gapsSep = IntStream.range(miG, maG).filter(i -> !gapsFull.contains(i)).boxed().collect(Collectors.toSet());
        }
        
        if (!gapsFull.isEmpty() && !gapsSep.isEmpty() && !gaps1.isEmpty())  limits = new int[] {miG, maG};
        else if (gapsFull.isEmpty() && full.size() < 3)                     limits = new int[] {miF, maF};
        else if (!gapsFull.isEmpty() && gapsSep.isEmpty())                  limits = gapsFull.size()<3 ?  new int[] {miG, maF} : new int[] {miG, maG};
        else                                                                limits = Arrays.copyOfRange( IntStream.range(miF, maF)
                                                                                                                  .filter( i -> !full .containsKey(i) || full.getOrDefault(i+1,0) > full.get(i)*1.5)
                                                                                                                  .toArray(), 0, 2);
        StringBuilder sb = new StringBuilder ();
        m = BITS_PATTERN.matcher(stripBits);
        while (m.find()) {
            String seq = m.group();
            sb.append( CONVERT_BITS.get(Arrays.asList(seq.charAt(0) == '1', seq.length() <= limits[0], seq.length() <= limits[1])) );
        }
        return sb.toString();
    }
    
    public static String decodeMorse(String morseCode) {
        if (morseCode.length() == 0) return "";
        return Arrays.stream( morseCode.split("   ") )
                     .map( word -> Arrays.stream( word.split(" ") ).map( ch -> MorseCode.get(ch) ).collect(Collectors.joining("")) )
                     .collect(Collectors.joining(" "));
    }
}

File: 2 kyu/Decode the Morse code, for real/solution.js
function classify(items, centroids) {
  var clusters = [];
  for (let i = 0; i < centroids.length; i++) {
    clusters.push([]);
  }

  items.forEach(item => {
    var distances = [];
    var cluster_index = 0;
    
    for (let i = 0; i < centroids.length; i++) {
      distances.push(Math.sqrt(Math.pow(item.length - centroids[i], 2)));
      //distances.push(Math.abs(item.length - centroids[i]));
    }

    for (let i = 0, min = Math.min(...distances); i < distances.length; i++) {
      if (distances[i] === min) {
         if (i === 2 && item[0] === '1') 
          cluster_index = 1;
         else
          cluster_index = i;
        break;
      };
    }

    clusters[cluster_index].push(item);
  });
  
  return clusters;
}

function mean(items) {
  if (!items.length) return 0;
  return items.reduce((sum, item) => sum + item.length, 0) / items.length;
}

function kmeans(data, centroids, iterations = Infinity) {
  var clusters = null, moved = false;
  
  do {
    clusters = classify(data, centroids);
    for (let i = 0, m; i < clusters.length; i++) {
      m = mean(clusters[i]);
      if (centroids[i] !== m) {
        centroids[i] = m;
        moved = true;
        continue;
      }
      moved = false;
    }
  }
  while (iterations-- && moved)

  return { clusters: clusters, centroids: centroids }
}

function decodeBitsAdvanced(bits){
    var map = {};
    var bits = bits.replace(/^0+|0+$/g, '').match(/1+|0+/g);
    if (!bits) return '';
    
    var result = kmeans(bits, [1,3,7], 100);
    var clusters = result.clusters; 
    var averages = [
      (Math.max(...clusters[0].map(item => item.length)) + Math.min(...clusters[0].map(item => item.length))) / 2,
      (Math.max(...clusters[1].map(item => item.length)) + Math.min(...clusters[1].map(item => item.length))) / 2,
      (Math.max(...clusters[2].map(item => item.length)) + Math.min(...clusters[2].map(item => item.length))) / 2,
    ]

    var centroids = [
      (averages[0] + averages[1]) / 2 || averages[0] || averages[1],
      (averages[1] + averages[2]) / 2 || averages[1] || averages[0] * 3,
    ]

    bits.reduce((map, signal) => {
      signal = signal.length;

      if (signal <= centroids[0]) {
        map['1'.repeat(signal)] = '.';
        map['0'.repeat(signal)] = '';
      }

      else if (signal <= centroids[1]) {
        map['1'.repeat(signal)] = '-';
        map['0'.repeat(signal)] = ' ';
      }

      else if (signal > centroids[1]) {
        map['1'.repeat(signal)] = '-';
        map['0'.repeat(signal)] = '   ';
      }

      return map;
    }, map)

    console.log(centroids, result.centroids);
    
    return bits.map(signal => {
      return map[signal];
    }).join('');
}

function decodeMorse(morseCode){
    if (!morseCode.length) return '';
    return morseCode.split('   ').map(word => {
      return word.trim().split(' ').map(code => {
        return MORSE_CODE[code] || code
      }).join('');
    }).join(' ');
}
_______________________________________________
var decodeBitsAdvanced = function(bits){
    bits = bits.replace(/(^0+|0+$)/g, "");
    if(!bits) return bits;
    if(!bits.replace(/1+/g, "")) return '.';
    
    let length = bits.match(/0+|1+/g).map(s=>s.length).sort((x, y)=> x-y);
    let [min, max] = [length.shift(), length.pop()];
    let unit = (max / 7 + min) / 2;
    let seven = Math.floor(7 * unit);
    let three = Math.round(3 * unit);
    
    return bits
            .replace(new RegExp("0".repeat(seven) + "+", "g"), "   ")
            .replace(new RegExp("1".repeat(three) + "+", "g"), "-")
            .replace(new RegExp("0".repeat(three) + "+", "g"), " ")
            .replace(/1+/g, ".")
            .replace(/0+/g, "")
}

var decodeMorse = function(morseCode){
   let words = morseCode.split("   ")
              .map(word => word.split(" ")
              .map(letter => MORSE_CODE[letter]))
              .map(letters => letters.join(""))
              .join(" ")
    return words;
}
_______________________________________________
var decodeBitsAdvanced = function(bits){
    // ToDo: Accept 0's and 1's, return dots, dashes and spaces
  if (!bits.includes('1')) return '';
  
  const arrayOfOnes = bits.replace(/^0+|0+$/g, '').split(/0+/).map(symb => symb.length);
  const arrayOfZeroes = bits.replace(/^0+|0+$/g, '').replace(/^1+|1+$/g, '').split(/1+/).map(symb => symb.length || 100); 
  const arrayOfAll = [...arrayOfOnes, ...arrayOfZeroes.filter(num => num <= Math.max(...arrayOfOnes) )];
  
  let dotCenter    = Math.min(...arrayOfAll),
      dashCenter   = Math.max(...arrayOfAll),
      spaceCenter  = dashCenter + 3,
      middleCenter = (dotCenter + dashCenter) / 2,
      prevDotCenter, prevDashCenter;
  
  while (prevDotCenter !== dotCenter || prevDashCenter !== dashCenter) {
    if (dotCenter === dashCenter) { 
      spaceCenter = 3 + (dashCenter = dotCenter * 3 - 1);
      break;
    };
    
    [prevDotCenter, prevDashCenter] = [dotCenter, dashCenter];
    let dotDistances = 0, dashDistances = 0, dots = 0, dashes = 0;
    arrayOfOnes.forEach( num => num <= middleCenter ? (dotDistances += num) && (dots += 1) : (dashDistances += num) && (dashes += 1));
    dotCenter = (dotDistances / dots) || dotCenter;
    dashCenter = (dashDistances / dashes);
    middleCenter = (dotCenter + dashCenter) / 2;
  }
  
  middleCenter = Math.floor((dotCenter + dashCenter) / 2 * 1.1) + 1;
  
  return bits.replace(new RegExp(`1{${middleCenter},}`, 'g'), '-')
             .replace(/1+/g, '.')
             .replace(new RegExp(`0{${spaceCenter},}`, 'g'), '  ')
             .replace(new RegExp(`0{${middleCenter},}`, 'g'), ' ')
             .replace(/0+/g, '');
}

var decodeMorse = function(morseCode){
    // ToDo: Accept dots, dashes and spaces, return human-readable message
  return morseCode.split('  ').map(word => word.split(' ').map(symb => MORSE_CODE[symb]).join('')).join(' ').trim();
}
_______________________________________________
var correctOneBit = function( unit, maxOneLength ) {
  if( maxOneLength < 3 ) return '1';

  let middle = maxOneLength / 2;
  if( middle < 2 ) middle = 2;

  if( unit.length <= middle ) return '1';
  else return '111';
};

var correctZeroBit = function( unit, maxZeroLength ) {
  if( maxZeroLength < 2 ) return '0';

  let divider = 2.4;
  if( maxZeroLength > 10 ) divider = 3.6;
  
  let lowerEdge = maxZeroLength / divider;
  if( lowerEdge < 2 ) lowerEdge = 2;

  let upperEdge = ( maxZeroLength - lowerEdge ) / 2 + lowerEdge;
  if( upperEdge < 6 ) upperEdge = 6;

  if( unit.length < lowerEdge ) return '0';
  else if( unit.length >= lowerEdge && unit.length < upperEdge ) return '000';
  else return '0000000';
};

var decodeBitsAdvanced = function( bits ) {
  bits = bits.replace( /(^0+|0+$)/g, '' );
  if( bits === '' ) return '';

  let minOneLength = 0;
  let maxOneLength = 0;
  let minZeroLength = 0;
  let maxZeroLength = 0;
  let parsedBits = bits.match( /0+|1+/g );
  let timeUnit = Math.min.apply( null, parsedBits.map( bit => bit.length ) );
  let parsedUnits = parsedBits.map( bit => {
    let value = bit[ 0 ].indexOf( '1' ) !== -1 ? '1' : '0';
    let shortedLength = Math.ceil( bit.length / timeUnit );

    if( value === '1' && !minOneLength ) minOneLength = shortedLength;
    if( value === '0' && !minZeroLength ) minZeroLength = shortedLength;

    if( value === '1' ) {
      maxOneLength = Math.max( maxOneLength, shortedLength );
      minOneLength = Math.min( minOneLength, shortedLength );
    }
    if( value === '0' ) {
      maxZeroLength = Math.max( maxZeroLength, shortedLength );
      minZeroLength = Math.min( minZeroLength, shortedLength );
    }

    return {
      value: value,
      length: shortedLength
    };
  } );
  
  let correctedBits = parsedUnits.map( unit => {
    if( !minZeroLength ) return '1';

    switch( unit.value ) {
      case( '1' ):
        return correctOneBit( unit, maxOneLength );
        break;
      default:
        return correctZeroBit( unit, maxZeroLength ); 
        break;
    }
  } ).join( '' );
    
  return correctedBits
    .replace( new RegExp( '111', 'g' ), '-' )
    .replace( new RegExp( '1', 'g' ), '.' )
    .replace( new RegExp( '0000000', 'g' ), '   ' )
    .replace( new RegExp( '000', 'g' ), ' ' )
    .replace( new RegExp( '0', 'g' ), '' );
};

var decodeMorse = function( morseCode ) {
    return morseCode.replace( /   /g, ' __ ' )
      .split( ' ' )
      .map( letter => letter === '__' ? ' ' : MORSE_CODE[ letter ] )
      .join( '' );
};
